File: backend/app/routers/season.py
import uuid

import httpx
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from ..deps import verify_user_session
from ..engine_client import engine_client

router = APIRouter(prefix="/season", tags=["시즌 (Season)"])


class ClaimRewardBody(BaseModel):
    level: int = Field(ge=1)
    track: str = Field(pattern=r"^(FREE|PREMIUM)$")


@router.get("/current")
async def get_current_season(
    _uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    try:
        return await engine_client.get_season_current()
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None


@router.get("/pass")
async def get_season_pass(
    uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    try:
        season = await engine_client.get_season_current()
        pass_data = await engine_client.get_season_pass(str(uid))
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None

    level = pass_data.get("current_level", 0)
    max_lvl = season.get("max_level", 30)
    sxp_per = season.get("sxp_per_level", 100)
    claimed = pass_data.get("claimed_levels", [])

    tier = "ROOKIE"
    if level >= 25:
        tier = "LEGEND"
    elif level >= 15:
        tier = "VETERAN"
    elif level >= 5:
        tier = "EXPLORER"

    reward_levels = [1, 5, 10, 15, 20, 25, 30]
    rewards = []
    for lv in reward_levels:
        if lv > max_lvl:
            break
        rewards.append(
            {
                "level": lv,
                "free_reward": {"type": "GOLD", "label": f"Gold {lv * 50}"},
                "premium_reward": {"type": "ITEM", "label": f"Item Lv.{lv}", "rarity": "R" if lv >= 15 else "E"},
                "is_claimed_free": lv in claimed,
                "is_claimed_premium": lv in claimed,
            }
        )

    return {
        "season": {
            "id": season["season_code"],
            "name": season["display_name"],
            "status": season["status"],
            "ends_at": season["ends_at"],
            "total_levels": max_lvl,
        },
        "current_level": level,
        "current_sxp": pass_data.get("sxp_balance", 0),
        "sxp_to_next": sxp_per,
        "is_premium": pass_data.get("has_premium", False),
        "tier_label": tier,
        "rewards": rewards,
    }


@router.get("/levels/{season_code}")
async def get_season_levels(
    season_code: str,
    uid: uuid.UUID = Depends(verify_user_session),
) -> list:
    try:
        return await engine_client.get_season_levels(season_code, str(uid))
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None


@router.post("/claim")
async def claim_season_reward(
    body: ClaimRewardBody,
    uid: uuid.UUID = Depends(verify_user_session),
) -> dict:
    try:
        return await engine_client.claim_season_reward(str(uid), body.level, body.track)
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code, detail=e.response.json().get("detail", str(e))
        ) from None

File: frontend/src/api/season.ts
import { USE_MOCK, api } from './client';
import type { ItemRarity } from './gacha';

export interface Season {
  id: string;
  name: string;
  status: 'ACTIVE' | 'UPCOMING' | 'ENDED';
  ends_at: string;
  total_levels: number;
  description?: string;
}

export interface SeasonRewardNode {
  level: number;
  free_reward: { type: 'GOLD' | 'XP' | 'ITEM' | 'BOX'; label: string; rarity?: ItemRarity };
  premium_reward?: { type: 'GOLD' | 'XP' | 'ITEM' | 'BOX'; label: string; rarity?: ItemRarity };
  is_claimed_free: boolean;
  is_claimed_premium: boolean;
}

export interface SeasonPass {
  season: Season;
  current_level: number;
  current_sxp: number;
  sxp_to_next: number;
  is_premium: boolean;
  tier_label: string;
  rewards: SeasonRewardNode[];
}

const MOCK_SEASON: SeasonPass = {
  season: {
    id: 'season-1',
    name: 'Tết Season Pass',
    status: 'ACTIVE',
    ends_at: new Date(Date.now() + 18 * 24 * 3600 * 1000 + 23 * 3600 * 1000).toISOString(),
    total_levels: 30,
  },
  current_level: 14,
  current_sxp: 1430,
  sxp_to_next: 1500,
  is_premium: false,
  tier_label: 'VETERAN',
  rewards: [
    { level: 1,  free_reward: { type: 'GOLD', label: 'Gold 50' },   premium_reward: { type: 'GOLD', label: 'Gold 100', rarity: 'E' }, is_claimed_free: true,  is_claimed_premium: true  },
    { level: 5,  free_reward: { type: 'BOX', label: 'C BOX' }, premium_reward: { type: 'ITEM', label: 'E ×1', rarity: 'E' }, is_claimed_free: true,  is_claimed_premium: true  },
    { level: 10, free_reward: { type: 'GOLD', label: 'Gold 200' },  premium_reward: { type: 'XP', label: 'XP 30', rarity: 'E' }, is_claimed_free: true,  is_claimed_premium: true  },
    { level: 14, free_reward: { type: 'ITEM', label: 'Claim', rarity: 'R' }, premium_reward: { type: 'ITEM', label: 'Claim', rarity: 'E' }, is_claimed_free: false, is_claimed_premium: false },
    { level: 15, free_reward: { type: 'ITEM', label: 'R ×1', rarity: 'R' }, premium_reward: { type: 'ITEM', label: 'L+Trail', rarity: 'L' }, is_claimed_free: false, is_claimed_premium: false },
    { level: 20, free_reward: { type: 'GOLD', label: 'Gold 500' },  premium_reward: { type: 'ITEM', label: 'L ×1', rarity: 'L' }, is_claimed_free: false, is_claimed_premium: false },
    { level: 25, free_reward: { type: 'XP', label: 'E ×1', rarity: 'E' }, premium_reward: { type: 'ITEM', label: 'M ×1 ⭐', rarity: 'M' }, is_claimed_free: false, is_claimed_premium: false },
    { level: 30, free_reward: { type: 'ITEM', label: 'L ×1', rarity: 'L' }, premium_reward: { type: 'ITEM', label: 'M ×1 ⭐⭐', rarity: 'M' }, is_claimed_free: false, is_claimed_premium: false },
  ],
};

export async function fetchSeasonPass(): Promise<SeasonPass> {
  if (USE_MOCK) return api.delay(MOCK_SEASON, 200);
  return api.realFetch<SeasonPass>('/season/pass');
}

export async function fetchCurrentSeason(): Promise<Season> {
  if (USE_MOCK) return api.delay(MOCK_SEASON.season, 150);
  return api.realFetch<Season>('/season/current');
}
